climakitae/threshold_tools.py

```diff
@@ -14,41 +14,179 @@
+            )
+        # In this case, identify the min (max) value of the grouped values for
+        # each window of length `grouped_duration``. Must be in `days`.
+        dur2_len, dur2_type = grouped_duration
+        if dur2_type != "day":
+            raise ValueError(
+                "`grouped_duration` specification must be in days. example: `grouped_duration = (3, 'day')`."
+            )
+
+        # Now select the min (max) from the duration period
+        if extremes_type == "max":
+            da_series = da_series.rolling(time=dur2_len, center=False).min("time")
+        elif extremes_type == "min":
+            da_series = da_series.rolling(time=dur2_len, center=False).max("time")
+
+    # Now select the most extreme value for each block in the series
     if extremes_type == "max":
-        ams = da.resample(time="A").max(keep_attrs=True)
-        ams.attrs["extreme value extraction method"] = "block maxima"
-        ams.attrs["extremes type"] = "maxima"
-        ams.attrs["block size"] = "1 year"
-        ams.attrs["timeseries type"] = "annual max series"
+        bms = da_series.resample(time=f"{block_size}A").max(keep_attrs=True)
+        bms.attrs["extremes type"] = "maxima"
+    elif extremes_type == "min":
+        bms = da_series.resample(time=f"{block_size}A").min(keep_attrs=True)
+        bms.attrs["extremes type"] = "minima"
+
+    # Calculate the effective sample size of the computed event type in all blocks, check the average value
+    if check_ess:
+        all_ess = []
+
+        # handle the ESS check differently depending on if it's spatial data
+        if "x" in da_series.dims and "y" in da_series.dims:
+            # case for spatial data, using stack/unstack to apply calculate_ess
+            stacked_da = (
+                da_series.stack(allpoints=["y", "x"]).dropna(dim="allpoints").squeeze()
+            )
+            for yr in set(bms.time.dt.year.values[0:-1]):
+                ess = xr.apply_ufunc(
+                    calculate_ess,
+                    stacked_da.sel(time=slice(f"{yr}", f"{yr+block_size-1}")).groupby(
+                        "allpoints"
+                    ),
+                    input_core_dims=[["time"]],
+                ).unstack("allpoints")
+                all_ess.append(ess)
+        else:
+            # case for not spacial data (no x, y dimensions)
+            for yr in set(bms.time.dt.year.values[0:-1]):
+                ess = calculate_ess(
+                    da_series.sel(time=slice(f"{yr}", f"{yr+block_size-1}"))
+                )
+                all_ess.append(ess)
+
+        average_ess = np.nanmean(all_ess)
+        if average_ess < 25:
+            print(
+                f"WARNING: The average effective sample size in your data is {round(average_ess, 2)} per block, which is low. This may result in biased estimates of extreme value distributions when calculating return values, periods, and probabilities from this data."
```

Comment:
Without making this a longer error, I think we should modify this somewhat to include the target of 25 (also, statistically, where did 25 come from?), and/or advice on how to increase the ESS, because a warning for example of "the average ESS in your data is 14.23 per block which is low" may mean very little to non-technical users as to how to fix (and maybe this technical user). It was not immediately obvious to me that I needed to pick a location with more grid cells/finer spatial resolution (for spatial xy data), and not increase the length of the time record being analyzed. 

I'm going to amend this, all my attempts to increase ESS have gotten me up to about 15, so what is the recommendation for increasing ESS (or above 25)?

Reply:
Noting that this is testing in threshold_basics notebook